internal/app/referenceclient/wire_details.go

```diff
@@ -206,12 +235,255 @@ func getBodyEndStream(trace tracer.Trace) (string, bool) {
+		return
+	}
+	msgFromDebug := msgType.New()
+	if err := protojson.Unmarshal(debugJSON, msgFromDebug.Interface()); err != nil {
+		// It is possible that the debug data is actually a google.protobuf.Any, and the "@type" property
+		// is causing the above unmarshal step to fail. So we'll try unmarshaling from Any next to see
+		// if that is successful.
+		//
+		// NOTE: this fallback logic doesn't really work if the actual message type is a
+		// google.protobuf.Value or google.protobuf.Struct. These types allow arbitrary JSON and will
+		// accept the data in the above step, even if it has the extra "@type" property. Luckily, that
+		// is not an issue since none of the conformance test cases try to use these types as error
+		// details.
+		//
+		// TODO: This fallback is mainly to accommodate the current connect-go implementation, which
+		//       includes the "@type" attribute because it just marshals the Any error detail message
+		//       for the debug value. But we'd prefer that attribute NOT be in the debug value, in which
+		//       case we could remove this fallback from here and let these checks be a bit more strict.
+		var anyMsg anypb.Any
+		if anyErr := protojson.Unmarshal(debugJSON, &anyMsg); anyErr != nil {
+			// It's not a valid Any either. So report the original error
+			printer.Printf("connect error JSON: details[%d]: could not unmarshal message %q from debug JSON: %v", i, msgName, err)
+			return
+		}
+		typeNameFromAny := anyMsg.TypeUrl[strings.LastIndexByte(anyMsg.TypeUrl, '/')+1:]
+		if typeNameFromAny != msgName {
+			printer.Printf("connect error JSON: details[%d]: debug data indicates type %q but should indicate type %q", i, typeNameFromAny, msgName)
+			return
+		}
+		msgFromAny, err := anyMsg.UnmarshalNew()
+		if err != nil {
+			printer.Printf("connect error JSON: details[%d]: could not unmarshal message %q from debug JSON: %v", i, msgName, err)
+			return
+		}
+		msgFromDebug = msgFromAny.ProtoReflect()
+	}
+	diff := cmp.Diff(msgFromValue.Interface(), msgFromDebug.Interface(), protocmp.Transform())
+	if diff != "" {
+		printer.Printf("connect error JSON: details[%d]: debug data does not match value: - value, + debug\n%s", i, diff)
+	}
+}
+
+func examineConnectEndStream(endStreamJSON json.RawMessage, printer internal.Printer) {
+	var endStream connectEndStream
+	if err := json.Unmarshal(endStreamJSON, &endStream); err != nil {
+		printer.Printf("connect end stream JSON is malformed: %v", err)
+		return
+	}
+	// Extra checks, since encoding/json above is lenient.
+	if _, err := checkNoDuplicateKeys("", json.NewDecoder(bytes.NewReader(endStreamJSON))); err != nil {
+		printer.Printf("connect end stream JSON: %v", err)
+		return
+	}
+	var asAny map[string]any
+	if err := json.Unmarshal(endStreamJSON, &asAny); err != nil {
+		// note: since above unmarshal step succeeded, this should never fail
+		printer.Printf("connect end stream JSON is malformed: %v", err)
+		return
+	}
```

Comment:
Is it worth moving out this logic into its own function? I noticed that `examineConnectError` and `examineConnectErrorDetail` perform very similar logic (albeit with some minor variations). All 3 seem to do a series of validation steps against a `json.RawMessage` and then use an `internal.Printer` to report an error. It may not even be worth it, but just an observation.

Reply:
Interesting point. I looked and was able to consolidate this into an `examineJSON` helper function. Just pushed a change with that.